clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java

```diff
@@ -497,14 +517,22 @@ public boolean hasNext() {
         @Override
         public Record next() {
             if (readRecords >= numRecords)
-                throw new NoSuchElementException();
+                throw new InvalidRecordException("Garbled data resulting in attempt of reading past the end of batch");
```

Comment:
Seems we can revert this change. We'll hit the InvalidRecordException below if there is additional data, so this is really more for iterator sanity.